core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -409,4 +417,50 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    private static class QueueSizeAdjuster
+    {
+        private final Map<String, Long> splitsWeightPerNode;
+        private final long minPendingSplitsWeightPerTask;
+        private final long maxPendingSplitsWeightPerTask;
+        private final long maxSplitsWeightPerNode;
+
+        QueueSizeAdjuster(long minPendingSplitsWeightPerTask, long maxPendingSplitsWeightPerTask, long maxSplitsWeightPerNode)
+        {
+            this.splitsWeightPerNode = new HashMap<>();
+            this.maxPendingSplitsWeightPerTask = maxPendingSplitsWeightPerTask;
+            this.minPendingSplitsWeightPerTask = minPendingSplitsWeightPerTask;
+            this.maxSplitsWeightPerNode = maxSplitsWeightPerNode;
+        }
+
+        public void update(List<RemoteTask> existingTasks, NodeAssignmentStats nodeAssignmentStats)
+        {
+            if (maxPendingSplitsWeightPerTask == minPendingSplitsWeightPerTask) {
+                return;
+            }
+            for (RemoteTask task : existingTasks) {
+                String nodeId = task.getNodeId();
+                long currentSplitsWeightPerNode = getMaxPendingSplitsWeightPerTask(nodeId);
+
+                if (currentSplitsWeightPerNode < maxPendingSplitsWeightPerTask && nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId) == 0) {
+                    currentSplitsWeightPerNode = Math.min(maxPendingSplitsWeightPerTask, currentSplitsWeightPerNode * 2);
+                }
+                else if (currentSplitsWeightPerNode > minPendingSplitsWeightPerTask && nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId) < (getMaxSplitsWeightPerNode(nodeId) / 3)) {
```

Comment:
I think in practice you can reduce it when:
```
(splits_queued_during_last_schedule - slits_scheduled_now) <= new_max_splits_per_node * static_coefficient;
```
e.g. we want to make sure that queue is big enough so that splits won't be drained to 0 before next schedule.

Reply:
Is this is not some kind not indirectly related to what it is right now? We test all pending splits against all possible assignments, with some margin. Adding tracking about last assignment could make this complex and under perform.

Reply:
discussed offline
```
current_queue_size=90
before: 90
after: 29

29 < current_queue_size / 3 => true

before - after = 61
adjust_queue_size => 90/2 => 45

61 > 45
```

but should be:
```
before: 80
after: 20

min_queue_size=60
```